sql/schemachanger: add support for CREATE SEQUENCE in declarative schema changer

fqazi opened the pull request:
- Adds support for statement phases that follow multiple transitions for descriptors that are in an adding state (additionally, allow limits on the maximum phase to which deprules will apply)
- Disables support for two version invariants for descriptors in an adding state
- Fixes prefix resolution creating new objects, where two-part names were not handled properly before
- Adds support for CREATE sequence in the declarative schema changer

fqazi committed:
sql/schemachanger: support multiple transitions during statement phase

Previously, the declarative schema changer, only allowed
a single element transition per statement phase, which would
limit the types of CREATE operations we could support. For
example for CREATE TABLE/SEQUENCE/VIEW we will decompose
these operations on columns and indexes, all of which
will execute in a stage. This patch, will add support
for more complex statement phases, by moving this enforcement
into the rules frame work. Additionally, the two version
invariant will not apply to cases when a descriptor is
being added.

Epic: None
Release note: None

fqazi committed:
sql/schemachanger: fix prefix resolution in declarative schema changer

Previously, when trying to resolve prefixes for newly
created objects, we didn't handle two parts names
properly in the declarative schema changer. This,
patch fixes this logic to match the resolver.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/104240)
<!-- Reviewable:end -->

fqazi committed:
sql/schemachanger: skip validation and backfill for added tables

Previously, we would always generate operations to backfill
and validate on newly created tables. This will prevent
us from being able to generate single stage plans for
creating relations. This patch removes those checks
for added tables.

Epic: none
Release note: None

fqazi committed:
sql/schemachanger: add support for the CREATE SEQUENCE

Previously, the CREATE SEQUENCE command only worked
in the legacy schema changer. This patch, adds support
for creating sequences in declarative schema changer.

Epic: None

Release note: None

postamar commented:
Nice work. Can you submit these as distinct PRs please? This is a bit unwieldy to review.

fqazi commented:
@postamar Let me split this into infrastructure work versus the actual CREATE SEQUENCE, which is smaller.
